### PR TITLE
[protobuf] fix protobuf-3.18 mingw build failure

### LINKED ISSUE #8992: protobuf 3.18.0 broken for Windows when building with MinGW-w64
protobuf 3.18.0 fails to build on Windows with MinGW-w64 GCC with the following errors:
```
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:108:28: error: 'F_GETFL' was not declared in this scope
  108 |   int flags = fcntl(file_, F_GETFL);
      |                            ^~~~~~~
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:108:15: error: 'fcntl' was not declared in this scope
  108 |   int flags = fcntl(file_, F_GETFL);
      |               ^~~~~
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:109:13: error: 'O_NONBLOCK' was not declared in this scope
  109 |   flags &= ~O_NONBLOCK;
      |             ^~~~~~~~~~
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:110:16: error: 'F_SETFL' was not declared in this scope
  110 |   fcntl(file_, F_SETFL, flags);
      |                ^~~~~~~
```

The solution is to replace `_MSC_VER` with `_WIN32` in `src/google/protobuf/io/zero_copy_stream_impl.cc`.

Apparently a false assumption was made `_MSC_VER` is always set on the Windows platform, but that is only the case when building with MSVC. On MSVC and MinGW gcc and clang `_WIN32` is always defined however.

### PR DESCRIPTION
**Pull request by fairlyblank**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes [this issue](https://github.com/protocolbuffers/protobuf/issues/8992)

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/protobuf/fix-mingw-build.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
++#ifndef _WIN32
+ #include <fcntl.h>
+ #include <sys/stat.h>
+ #include <sys/types.h>
+@@ -104,7 +104,7 @@
+       is_closed_(false),
+       errno_(0),
+       previous_seek_failed_(false) {
+-#ifndef _MSC_VER
++#ifndef _WIN32
```
Comment (JackBoosY) [inline ports/protobuf/fix-mingw-build.patch:17]: I think this should wait for the upstream approval, can you please make a PR to the upstream first?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/protobuf/fix-mingw-build.patch (hunk 1 above)
Comment (fairlyblank) [reply ports/protobuf/fix-mingw-build.patch:17]: It seems like that was fixed by [this commit](https://github.com/protocolbuffers/protobuf/commit/15c63ad850c0b60a728414fd107c5ed3b3eb2772), tagged as version 3.19.*

Comment (JackBoosY) [reply ports/protobuf/fix-mingw-build.patch:17]: So the first place didn't change to `#ifndef _WIN32`, can you revert that?

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Can you please check the mingW build on your machine?

**Comment by fairlyblank** [comment]
> Can you please check the mingW build on your machine?

it works fine

**Comment by strega-nil-ms** [comment]
Thanks!

**Pull request merged**